### PR TITLE
Implement more RESTful routes

### LINKED ISSUE #115: More RESTful style endpoints
In #106 @Bluenix2 suggested the below. I really like this proposed change, so I think we should do it after 106 is merged.

I am not sure how thoroughly this has been discussed, and how much of a say I have in this. But I have some opinions on the new endpoints, you currently have:

GET /size: Get the size of the canvas
GET /pixels: Get the whole canvas
GET /pixel: Get a specific pixel
PUT /pixel: Set a specific pixel
One quick thing I want to mention about /pixels, it doesn't really make sense, we're already at pixels.pythondiscord.com. Making the URL pixels.pythondiscord.com/pixels..

In my opinion, the following seem more RESTful:

GET /canvas/size: Get the size of the canvas alternatively this info could be included in the headers and would be mainly requested using HEAD /canvas.
GET /canvas: Get the whole canvas
alternatively the following two endpoints could use /canvas/pixels
GET /canvas?x=x&y=y: Get a specific pixel
PUT /canvas: Set a specific pixel you could also use query parameters here and the body would only consist of the color

### PR DESCRIPTION
**Pull request by Bluenix2**
I am starting this pull request as a draft so that there's no conflicts between work. Resolves #115.

So I've thought about these a bit more, and it would probably be better to do the following..
- GET /canvas/                              *Get the size; this should be the first one the user tries, that's why it's the simplest*
- GET /canvas/pixels                     *Get all pixels*
- GET /canvas/pixels?x=0&y=0    *Get a specific pixel*
- PUT                                             *Set a specific pixel, unsure which one is best*
  - /canvas/pixels    *x, y and rgb is in the body*
  - /canvas/pixels?x=0&y=0    *only rgb in the body, but conforms better to how you get a pixel*

The biggest issue with this *second proposal* is that now we have a "conflict" of `GET /canvas/pixels` and `GET/canvas/pixels?x=0&y=0`. We cannot have 2 documentation entries to outline this (if we use one that automatically generates documentation from the openapi spec).
> Query string parameters must not be included in paths. They should be defined as query parameters instead.
> This also means that it is impossible to have multiple paths that differ only in query string, such as:
> ```
> GET /users?firstName=value&lastName=value
> GET /users?role=value
> ```
> This is because Swagger considers a unique operation as a combination of a path and the HTTP method, and additional parameters do not make the operation unique.

Of course it's not impossible, we can have one shared function. But this means we need to merge `get_pixel` and `pixels` doc-string. Leading to more confusion than UX we get.

That makes the first proposal conflict free, that said there may be improvements to be made?
- **GET /canvas/size**
- **GET /canvas/**
- GET /canvas/pixels?x=0&y=0
- PUT
  - /canvas/pixels
  - /canvas/pixels?x=0&y=0

### CODE REVIEW

**Comment by jb3** [comment]
Feedback on endpoints:
- Size should explicitly be in the endpoint, so something like `GET /canvas/size`.
- Fetching an individual pixel should not use pixels plural imo, I'd rather keep it as `GET /canvas/pixel?x=1&y=1` or put it in the URL parameters like `GET /canvas/pixels/1/2` and document the order.
- I don't like PUT with URL parameter *and* a body, it doesn't make sense. The document being created should be complete in the body, not partials in two different places.

**Comment by Bluenix2** [comment]
> * Fetching an individual pixel should not use pixels plural imo, I'd rather keep it as `GET /canvas/pixel?x=1&y=1` or put it in the URL parameters like `GET /canvas/pixels/1/2` and document the order.

Ah, interesting. The way I see it is that we're making a GET requests to get the pixels, but we're picking a specific one (using query parameters to narrow down). But yes using `GET /canvas/pixel?x=0&y=0` and `GET /canvas/pixelS` (plurals) would remove that conflict I was talking about.

> I don't like PUT with URL parameter and a body, it doesn't make sense. The document being created should be complete in the body, not partials in two different places.

Yeah the former would be better and then use the singular form as we do getting a specific pixel.

**Comment by Bluenix2** [comment]
While this is being done, should moderation endpoints also get similar changes?

**Comment by Bluenix2** [comment]
I have now done some basic testing and squashed the commits so that this can be rebased.

File: pixels/endpoints/general.py [hunk 1]
```diff
@@ -3,14 +3,44 @@
...
-from pixels.models import Message, Pixel
+from pixels.models import GetSize, Message, Pixel
 from pixels.utils import auth, ratelimits
 
 log = logging.getLogger(__name__)
-router = APIRouter(tags=["Canvas Endpoints"], dependencies=[Depends(auth.JWTBearer())])
 
+# By only adding the JWT dependency to the submounted router we can have size() be un-authenticated
+router = APIRouter(prefix="/canvas", tags=["Canvas Endpoints"])
+secure = APIRouter(dependencies=[Depends(auth.JWTBearer())])  # We include this at the bottom
```
Comment (ChrisLovering) [inline pixels/endpoints/general.py:13]:
```suggestion
# We include this at the bottom
secure = APIRouter(dependencies=[Depends(auth.JWTBearer())], tags=["Canvas Endpoints"])
```

**Review by ChrisLovering — CHANGES_REQUESTED** [review]

**Review by ChrisLovering — APPROVED** [review]

**Pull request merged**